### PR TITLE
Build rest-server from source

### PR DESCRIPTION
**Pull request by mmorhun**
Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

Build rest-server binary from source code instead of downloading official binary.
Add license file near the binary.

### CODE REVIEW

**Comment by nickboldt** [comment]
instead of two dockerfiles w/ runtime = alpine or ubi8-minimal, how about one dockerfile that uses go-toolset as builder and ubi8/ubi-micro (or scratch) as runtime image?

File: rhel.Dockerfile [hunk 1]
```diff
@@ -6,27 +6,26 @@
...
-    curl -sLO https://github.com/restic/rest-server/releases/download/v${REST_SERVER_VERSION}/rest-server_${REST_SERVER_VERSION}_linux_${ARCH}.tar.gz && \
-    tar -xzf rest-server_${REST_SERVER_VERSION}_linux_${ARCH}.tar.gz && \
-    cp rest-server_${REST_SERVER_VERSION}_linux_${ARCH}/rest-server /usr/local/bin/rest-server && \
-    rm -rf rest-server_${REST_SERVER_VERSION}_linux_${ARCH} rest-server_${REST_SERVER_VERSION}_linux_${ARCH}.tar.gz && \
-    microdnf remove -y tar gzip && \
-    microdnf clean all
+    mkdir -p $GOPATH && cd $GOPATH && \
+    git clone --depth 1 --branch $REST_SERVER_TAG https://github.com/restic/rest-server.git && \
+    cd rest-server && \
+    go mod vendor && \
```
Comment (nickboldt) [inline rhel.Dockerfile]: to downstream this, we would just have to remove the `git clone` and replace with a `COPY asset-rest-server.tgz` + unpack step, so +1 for this approach.

File: rhel.Dockerfile [hunk 2]
```diff
@@ -6,27 +6,26 @@
...
+    GOOS=linux GOARCH=${ARCH} CGO_ENABLED=0 go build -mod=vendor -o rest-server  ./cmd/rest-server
+
+# https://access.redhat.com/containers/?tab=tags#/registry.access.redhat.com/ubi8-minimal
+FROM registry.access.redhat.com/ubi8-minimal:8.4-200
```
Comment (nickboldt) [inline rhel.Dockerfile]: latest tag is 8.4-205, but if you want to switch to `ubi8/ubi-micro:8.4-72` (note: NOT ubi8-micro), that works too.

Could also use `FROM scratch` if you also copy the sleep binary into the container from the KIP build, like this 
* https://github.com/eclipse-che/che-machine-exec/blob/main/build/dockerfiles/Dockerfile#L16
* https://github.com/eclipse-che/che-machine-exec/blob/main/build/dockerfiles/Dockerfile#L60

If using micro, we could merge/remove the alpine based build and only have the go-toolset/micro one.

File: Dockerfile [hunk 3]
```diff
@@ -6,26 +6,26 @@
...
+    git clone --depth 1 --branch $REST_SERVER_TAG https://github.com/restic/rest-server.git && \
+    cd rest-server && \
+    go mod vendor && \
+    GOOS=linux GOARCH=${ARCH} CGO_ENABLED=0 go build -mod=vendor -o rest-server  ./cmd/rest-server
+
+# https://access.redhat.com/containers/?tab=tags#/registry.access.redhat.com/ubi8/ubi-micro
+FROM registry.access.redhat.com/ubi8/ubi-micro:8.4-72
+
+COPY --from=builder /tmp/go/rest-server/rest-server /usr/local/bin/rest-server
+COPY --from=builder /tmp/go/rest-server/LICENSE /usr/local/bin/rest-server-LICENSE.txt
```
Comment (benoitf) [inline Dockerfile:28]: do we need to copy the license in the image ?

File: Dockerfile [hunk 4]
```diff
@@ -6,26 +6,26 @@
 # SPDX-License-Identifier: EPL-2.0
 #
 
-FROM alpine:3.13.5
-
-ARG KUBERNETES_VERSION=1.21.1
-ARG REST_SERVER_VERSION=0.10.0
+# https://access.redhat.com/containers/?tab=tags#/registry.access.redhat.com/ubi8/go-toolset
+FROM registry.access.redhat.com/ubi8/go-toolset:1.15.7-11 as builder
```
Comment (benoitf) [inline Dockerfile]:
```suggestion
FROM registry.access.redhat.com/ubi8/go-toolset:1.15.13-4 as builder
```

File: Dockerfile [hunk 5]
```diff
@@ -6,26 +6,26 @@
...
-    curl -sLO https://github.com/restic/rest-server/releases/download/v${REST_SERVER_VERSION}/rest-server_${REST_SERVER_VERSION}_linux_${ARCH}.tar.gz && \
-    tar -xzf rest-server_${REST_SERVER_VERSION}_linux_${ARCH}.tar.gz && \
-    cp rest-server_${REST_SERVER_VERSION}_linux_${ARCH}/rest-server /usr/local/bin/rest-server && \
-    rm -rf rest-server_${REST_SERVER_VERSION}_linux_${ARCH} rest-server_${REST_SERVER_VERSION}_linux_${ARCH}.tar.gz && \
-    apk del curl
+    mkdir -p $GOPATH && cd $GOPATH && \
+    git clone --depth 1 --branch $REST_SERVER_TAG https://github.com/restic/rest-server.git && \
+    cd rest-server && \
+    go mod vendor && \
+    GOOS=linux GOARCH=${ARCH} CGO_ENABLED=0 go build -mod=vendor -o rest-server  ./cmd/rest-server
```
Comment (benoitf) [inline Dockerfile]:
```suggestion
    GOOS=linux GOARCH=${ARCH} CGO_ENABLED=0 go build -mod=vendor -o rest-server ./cmd/rest-server
```

**Review by benoitf — APPROVED** [review]

File: Dockerfile (hunk 3 above)
Comment (mmorhun) [reply Dockerfile:28]: Yes, this is the requirement, see https://github.com/eclipse/che/issues/20028

**Comment by benoitf** [comment]
Also FYI the PR build is failing

**Comment by mmorhun** [comment]
@benoitf I know, but it succeeds locally (however I am not using `buildx`) and seems this is a bug: https://github.com/docker/buildx/issues/101

**Review by AndrienkoAleksandr — APPROVED** [review]

**Pull request merged**